<a href="https://colab.research.google.com/github/dhdan97/CSC-488-Coursework/blob/main/a3_Daniel_H_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/dhdan97/CSC-488-Coursework/blob/main/a3_Daniel_H_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

\(\^Be sure to update this button to point to your notebook instead of the sample notebook\)

In [1]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
Slime = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
# Output the first 15 rows of the data
print(Slime.head(15))
# Display a summary of the table information (number of datapoints, etc.)
Slime.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = Slime[['Temperature °C', 'Mols KCL']]
y = Slime['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=42)

## Part 3. Perform a Linear Regression

In [4]:
import warnings

# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = [[469, 647]]
with warnings.catch_warnings(record=True): samplePrediction = str("{:#.5g}".format((reg.predict(sample)[0])))
print("With Temperature of 469°C and 647 Mols of KCL, our model predicts a size of " + samplePrediction + " nm^3")
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print("Given our test features and labels, our score is: " + str("{:#.5g}".format(reg.score(X_test,y_test))))

With Temperature of 469°C and 647 Mols of KCL, our model predicts a size of 6.6498e+05 nm^3
Given our test features and labels, our score is: 0.85525


###Our score is defined as (1 - u/v), where u is the residual sum of squares and v is the total sum of squares.

###This means our score is a ratio of how close our predicted answers are to the observed answers(labels) over how much our label values vary.

###The smaller our residual sum of squares and/or the greater our label variance, the better our score.

###I would consider a score of 0.85525 very good for our first result

In [5]:
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
reg.coef_

array([ 866.14641337, 1032.69506649])

In [6]:
reg.intercept_

-409391.47958340764

Sample equation: $E = mc^2$

$h(x) = 866.14x + 1032.6x - 4.0939 * 10^{5}$

## Part 4. Use Cross Validation

In [7]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(reg, X, y)
# Report on their finding and their significance
scores

array([0.83918826, 0.87051239, 0.85871066, 0.87202623, 0.84364641])

###With 5-fold cross validation, it looks like our 4th entry produced the greatest score. This means that the combination of spliting our data into 5 parts, and using the 4th part as our testing data gave us a greater score than any other combination of 5 parts here.

## Part 5. Using Polynomial Regression

In [8]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
polyX = poly.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(polyX, y, test_size=0.10, random_state=42)

reg = LinearRegression().fit(X_train, y_train)

In [9]:
# Report on the metrics and output the resultant equation as you did in Part 3.
print("Given our test features and labels, our score is: " + str("{:#.5g}".format(reg.score(X_test,y_test))))

Given our test features and labels, our score is: 1.0000


###A score of 1.0 means our model is perfect, i.e our residual sum of squares is 0.

In [10]:
reg.coef_

array([ 0.00000000e+00,  1.20000000e+01, -1.27195488e-07,  1.26494371e-11,
        2.00000000e+00,  2.85714287e-02])

In [11]:
reg.intercept_

2.0477105863392353e-05

$h(x) = 12x_{1} + (-1.2718*10^{7})x_{2} + (1.2649*10^{-11})x_{1}x_{2} + 2x_{1}^{2} + (2.8571*10^{-2})x_{2}^{2}$